torch/csrc/autograd/functions/accumulate_grad.cpp

```diff
@@ -38,6 +37,7 @@ auto AccumulateGrad::apply(variable_list&& grads) -> variable_list {
     new_grad = (*hook)({new_grad})[0];
   }
 
+  std::lock_guard<std::mutex> lock(grad_mutex);
```

Comment:
This is not needed in the case where there is only 1 thread per device in the autograd engine. But I guess it's not a big performance penalty so not worth the complexity to disable it in that case?